### DISCUSSION TITLE
Found somthing that's either a bug or a troublesome quirk

### DISCUSSION THREAD

**Discussion by mscout1** [wagtail/wagtail#10716] [General]
I ran into a problem updating from 3.0.3 to 4.0 (I am headed to 5.0, but I am taking my upgrades step-wise.)

[I asked about it on StackOverflow](https://stackoverflow.com/questions/76765355/in-wagtail-4-0-how-do-i-query-for-revisions-whose-pages-are-not-live), but it's had no attention so far. 

Even if this is intended behavior, it needs a better error. A sql exception from the db is a little rough.

**Comment by laymonage** [comment]
I've posted an answer on SO, copying it here for visibility:

If you're using PostgreSQL, this is caused by a 12-year-old Django issue https://code.djangoproject.com/ticket/16055 that will be fixed in Django 5.0 (yet to be released at the time of writing).

A possible workaround:

```python
from django.contrib.contenttypes.models import ContentType
from wagtail.models import Page, Revision

base_page_ct = ContentType.objects.get_for_model(Page)
not_live_ids = Page.objects.not_live().values_list('id', flat=True)
Revision.page_revisions.filter(base_content_type=base_page_ct, object_id__in=not_live_ids)
```

> How can I query for relations, filtering by a field on the related Page?

I assume you mean "query for revisions"?

Using a similar trick as above, you can filter the page first, use `values_list` to get the IDs, then filter the revisions by the base Page content type and the object IDs.

Since QuerySets are lazy, the first query to get the page IDs won't be evaluated directly and instead will be used as a subquery to revision query.

**Reply by mscout1** [reply]
Ah, yes. I thought it smelled like a bug, but I thought it was wagtail that was the bad guy. I didn't realize that mismatched types are actually supposed to work. Your work around is basically the same as the one that I came up with. (No one on the team was happy with it, which is why I was still pursuing the issue. )

Thanks for the help.
